src/ray/core_worker/reference_count.cc

```diff
@@ -553,6 +553,7 @@ void ReferenceCounter::UpdateObjectPinnedAtRaylet(const ObjectID &object_id,
     RAY_CHECK(it->second.owned_by_us);
     if (!it->second.OutOfScope(lineage_pinning_enabled_)) {
       it->second.pinned_at_raylet_id = raylet_id;
+      it->second.locations.emplace(raylet_id);
```

Comment:
Since these pinned location updates can happen before the Plasma-triggered location reports, I opted to eagerly add them to the object's locations. The later `AddObjectLocation` call will be a no-op since the location will already exist in the set.